src/ByteBuffer.cpp

```diff
@@ -771,10 +771,10 @@ bool cByteBuffer::WriteXYZPosition64(Int32 a_BlockX, Int32 a_BlockY, Int32 a_Blo
 {
 	CHECK_THREAD
 	CheckValid();
-	return WriteBEInt64(
-		(static_cast<Int64>(a_BlockX & 0x3FFFFFF) << 38) |
-		(static_cast<Int64>(a_BlockY & 0xFFF) << 26) |
-		(static_cast<Int64>(a_BlockZ & 0x3FFFFFF))
+	return WriteBEUInt64(
+		((static_cast<UInt64>(a_BlockX) & 0x3FFFFFF) << 38) |
+		((static_cast<UInt64>(a_BlockY) & 0xFFF) << 26) |
+		(static_cast<UInt64>(a_BlockZ) & 0x3FFFFFF)
```

Comment:
This just switches an unambiguously valid and defined left-shift to a confusing cast of a a signed integer to unsigned - although the result will be the right value it reles on befuddling semantics of casting negative integers to unsigneds.

Reply:
What if Int32 a_BlockX was a sign-magnitude representation? The protocol needs 2s complement! :o

[from here](https://timsong-cpp.github.io/cppwp/n4659/conv.integral)
> If the destination type is unsigned, the resulting value is the least unsigned integer congruent to the source integer (modulo 2n where n is the number of bits used to represent the unsigned type). [ Note: In a two's complement representation, this conversion is conceptual and there is no change in the bit pattern (if there is no truncation).  — end note ]

So it ensures a_BlockX is in 2s complement representation.

nb this is an exercise in amusement, not bugfixing

Reply:
Luckily C++20 mandates 2s complement, we don't have to worry about sign-magnitude anymore :)

Reply:
aye. nonetheless since we're interpreting a value and writing it out into serialised form as a sequence of bytes a u64 is appropriate for that.